src/mem/cache/prefetch/indirect_memory_v2.hh

```diff
@@ -0,0 +1,662 @@
+ * section 3.3.2 of the paper
+ */
+
+#ifndef __MEM_CACHE_IMP_HH__
+#define __MEM_CACHE_IMP_HH__
+
+#include <map>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "mem/cache/prefetch/queued.hh"
+#include "mem/packet.hh"
+#include "mem/packet_access.hh"
+#include "params/IMPv2Prefetcher.hh"
+#include "sim/system.hh"
+
+namespace gem5
+{
+    namespace prefetch
+    {
+        namespace IMPv2Internals
+        {
+            template<typename T>
+            void saturating_increment(T& old_value, const T& upper_bound)
+            {
+                if (old_value < upper_bound)
+                    old_value += 1;
+            }
+
+            template<typename T>
+            void saturating_decrement(T& old_value)
+            {
+                if (old_value > 0)
+                    old_value -= 1;
+            }
+
+            /*
+             * The StreamDetector performs two functions:
+             *  1. It detects stream accesses to the indexing array, B
+             *  2. If it is confident about a stream, it returns the predicted
+             *     delta for that stream so that IMP may issue stream
+             *     prefetches using the delta
+             */
+            class StreamDetector
+            {
+                private:
+                    struct StreamTableEntry
+                    {
+                        std::optional<Addr> previous_address;
+                        std::optional<Addr> delta;
+                        uint64_t confidence;
+
+                        StreamTableEntry()
+                        {
+                            previous_address = {};
+                            delta = {};
+                            confidence = 0;
+                        }
+                    };
+
+                    /*
+                     * The confidence threshold above which a stream is
+                     * classified as detected
+                     */
+                    const uint64_t streaming_threshold;
+
+                    /*
+                     * The stream table is indexed using the IP of the load
+                     * instruction
+                     */
+                    std::map<Addr, StreamTableEntry> stream_table;
+                    const uint64_t stream_table_size;
+
+                public:
+                    StreamDetector(std::size_t size, uint64_t threshold):
+                        streaming_threshold(threshold), stream_table_size(size)
+                    {}
+
+                    /*
+                     * Add a load address, addr, to the stream table.
+                     * The address was generated by a instruction which the IP
+                     * points to. Return the address delta if the confidence
+                     * is high enough for the stream associated with IP
+                     */
+                    std::optional<Addr> addToStreamTable(const Addr IP,
```

Comment:
This function adds an address to the table. I don´t think it makes sense for it to return an optional delta, as this is not related to the insertion itself. Instead, I think that this function should return void, and that inStreamTable´s return value should be changed so that it returns the entry.

Reply:
The `addToStreamTable` function performs two jobs:
1. It adds an entry to the stream table
2. It checks if the added entry's confidence was high enough and if it was, it returns the associated stride delta.
However, I do agree that this function can be renamed to indicate the two things that it does. Unfortunately I can't think of any concise name for this function which would encapsulate both these tasks, Do you have any suggestions?

The `inStreamTable` function does not modify the state of the Stream Detector. All that this function does is to answer whether or not a given address was predicted by the Stream Detector. Since we don't want any other information about the Stream Table entries from this function, I chose not to return a pointer to the entry, but rather simply return a boolean. This function is then used by the main `IMP` class to classify accesses as index accesses and indirected accesses. Index accesses, i.e. access to the array B if we are tracking A[B[i]], should get the `true` from this function, whereas accesses to the indirect array A should get `false`.

Reply:
For me that is exactly the issue: it performs two jobs. An alternative would be to add one or two more functions to do the other jobs (e.g., one to check if the entry with the given addr has a high-enough confidence, and the other to get the entry´s delta). The goal is to make the code more easily understandable, at the cost of slightly worse performance (due to having to do an extra lookup).